// rescript-test/test/BasicExample.mjs
// Generated by ReScript, PLEASE EDIT WITH CARE

import * as Test from "../src/Test.mjs";

function intEqual(message, a, b) {
  Test.assertion(message, "intEqual", (function (a, b) {
          return a === b;
        }), a, b);
}

function stringEqual(message, a, b) {
  Test.assertion(message, "stringEqual", (function (a, b) {
          return a === b;
        }), a, b);
}

Test.test("Equals", (function (param) {
        intEqual(undefined, 1, 1);
      }));

function isCharCode(a, b) {
  return a.charCodeAt(0) === b;
}

Test.test("Custom comparator", (function (param) {
        var a = "a";
        Test.assertion("Char code should match", "isCharCode", isCharCode, a, 97.0);
        Test.assertion("Char code should match", "isCharCode", isCharCode, a, 98.0);
      }));

function userEq(a, b) {
  return a.id === b.id;
}

function userEqual(message, a, b) {
  Test.assertion(message, "userEqual", userEq, a, b);
}

Test.test("DeepEquals", (function (param) {
        stringEqual(undefined, "user", "user");
        userEqual(undefined, {
              username: "user",
              id: "a"
            }, {
              username: "user",
              id: "a"
            });
      }));

export {
  intEqual ,
  stringEqual ,
  isCharCode ,
  userEq ,
  userEqual ,
}
/*  Not a pure module */
